x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,378 @@
+// 4. Reports collected metrics through the mb.ReporterV2
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	endDate := time.Now().UTC()
+
+	if !m.config.Collection.Realtime {
+		// If we're not collecting realtime data, then just pull until
+		// yesterday (in UTC).
+		endDate = endDate.AddDate(0, 0, -1)
+	}
+
+	startDate := endDate.AddDate(0, 0, -m.config.Collection.LookbackDays)
+
+	m.report = report
+	return m.fetchDateRange(startDate, endDate, m.httpClient)
+}
+
+// fetchDateRange retrieves OpenAI API usage data for each configured API key within a date range.
+//
+// For each API key:
+// 1. Retrieves last processed date from state store
+// 2. Adjusts collection range to avoid duplicates
+// 3. Collects daily usage data
+// 4. Updates state store with latest processed date
+// 5. Handles errors per day without failing entire range
+func (m *MetricSet) fetchDateRange(startDate, endDate time.Time, httpClient *RLHTTPClient) error {
+	for _, apiKey := range m.config.APIKeys {
+		// stateKey using stateManager's key prefix and hashing apiKey
+		stateKey := m.stateManager.GetStateKey(apiKey.Key)
+
+		lastProcessedDate, err := m.stateManager.GetLastProcessedDate(apiKey.Key)
+		if err == nil {
+			// We have previous state, adjust start date
+			startDate = lastProcessedDate.AddDate(0, 0, 1)
+			if startDate.After(endDate) {
+				continue
+			}
+		}
+
+		for d := startDate; !d.After(endDate); d = d.AddDate(0, 0, 1) {
+			dateStr := d.Format("2006-01-02")
+			if err := m.fetchSingleDay(dateStr, apiKey.Key, httpClient); err != nil {
+				m.logger.Errorf("Error fetching data for date %s: %v", dateStr, err)
+				continue
+			}
+			if err := m.stateManager.store.Put(stateKey, dateStr); err != nil {
+				m.logger.Errorf("Error storing state for API key: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// fetchSingleDay retrieves usage data for a specific date and API key.
+func (m *MetricSet) fetchSingleDay(dateStr, apiKey string, httpClient *RLHTTPClient) error {
+	req, err := m.createRequest(dateStr, apiKey)
+	if err != nil {
+		return fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		var netErr net.Error
+		if errors.As(err, &netErr) && netErr.Timeout() {
+			return fmt.Errorf("request timed out with configured timeout: %v and error: %w", m.config.Timeout, err)
+		}
+		return fmt.Errorf("error making request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error response from API: status=%s", resp.Status)
+	}
+
+	return m.processResponse(resp, dateStr)
+}
+
+// createRequest builds an HTTP request for the OpenAI usage API.
+func (m *MetricSet) createRequest(dateStr, apiKey string) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, m.config.APIURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	q := req.URL.Query()
+	q.Add("date", dateStr)
```

Comment:
What if we use [Set](https://pkg.go.dev/net/http#Header.Set) method here? It's depends on if you intend to append vs replace, and to avoid duplicate headers. If you decide to use Set, consider making that change down below as well.

Reply:
This is not headers. Adding query param here.

```
<api-endpoint>?date=<date>
```

`Add` is intentional here. User might supply us with ApiURL that has already some query params (intentionally), so we want to append to that.